### ISSUE TITLE
CBOR Sequences

### ISSUE THREAD

**Issue by ektrah** [core-wg/coral#43]
Should we use [CBOR Sequences](https://tools.ietf.org/html/draft-ietf-cbor-sequence-02) for CoRAL documents and/or Constrained Resource Identifiers?

**Comment by jimsch** [comment]
No.  The one byte small encoding is not worth the benefit of being able to encode the root of the document as a sequence.  You cannot include a sequence in the middle of a document without some type of wrapping.

**Comment by ektrah** [comment]
> No. The one byte small encoding is not worth the benefit of being able to encode the root of the document as a sequence.

Why did we waste one of the few precious content format IDs on `application/cbor-seq` then?

> You cannot include a sequence in the middle of a document without some type of wrapping.

I was thinking of wrapping a CoRI in a byte string rather than an array. That would allow a processor to determine the length of a CoRI in bytes for comparing it to another CoRI without decoding it...

**Comment by jimsch** [comment]
For the first - probably because it is Carsten's baby.  However, that is setup so that it will always be a top level sequence so you can save the first byte without any problems.   This is not a nested case.   There is a byte string wrapper for multipart so a sequence can be used there.  

Wrapping the CoRI in a byte string does allow for direct comparison, if some restrictions are placed on the encoding (i.e. minimum lengths).  Using a byte string rather than an array will make the final encoding longer.

**Comment by ektrah** [comment]
Yes, we could save 1 or 2 bytes by turning the top-level array into a CBOR sequence:

```diff
-document = body
+document = *(link / form / representation / directive)   ; CBOR Sequence

 body = [*(link / form / representation / directive)]
```

The remaining document would be unaffected.

This is a very minor issue, but I think the solution is somewhat attractive, because it's easy to implement and an empty document is actually encoded as a zero-length representation.

Let's use #2 to discuss using a CBOR sequence instead of an array for CoRIs.

**Comment by ektrah** [comment]
The con is that the CoRAL media type would be `application/coral+cborseq`.

**Comment by chrysn** [comment]
Let's do that; it does save a byte on the wire, and the only actual issues are the longer media type (so what, it'll get a code nonetheless) and possibly CBOR implementations that can't sequence yet -- for now this'll be an opportunity for them to add support, and if it turns out some implementers just can't do it on their platform, there's ample time to go back.

**Comment by ektrah** [comment]
It seems cbor.me doesn’t have support for decoding (a whole) CBOR sequence yet, which is a major blocker for CoRAL documents as sequences... @cabo

**Comment by jimsch** [comment]
I would rather save the code size here than the one byte on the wire.  You have just made the code for parsing the top level document and the nested set of link attributes different.

**Comment by ektrah** [comment]
From a mailing list discussion:

> The idea behind this is that a processor should be able to skip over an entire sub-tree without having to process all the directives in it. Before CoRAL used CBOR, each sub-tree used to be prefixed by its number of bytes rather than its number of data items, so skipping over a sub-tree could be implemented by simply bumping a pointer and doing a bounds check.

What if we changed *everything* to a CBOR sequence? (i.e., save one byte at the top-level and replace every array of nested elements/form fields with a `bstr .cborseq X`)

(Not sure about this; just thinking out loud...)

**Comment by cabo** [comment]
On 2020-01-12, at 18:50, ektrah <notifications@github.com> wrote:
> 
> What if we changed everything to a CBOR sequence? (i.e., save one byte at the top-level and replace every array of nested elements/form fields with a bstr .cborseq X)

Not sure how important skip-ability is here, but that would indeed work (at the expense of returning from counting elements to counting bytes, like ASN.1 did).

If this is only about code size, remember that bracketing a CBOR sequence with 0x9f/0xff turns it into a CBOR array.

Grüße, Carsten

**Comment by chrysn** [comment]
> You have just made the code for parsing the top level document and the
> nested set of link attributes different.  

Does it? I'd say it just changes your the file initialization from
"expect an element" to "you're already in an indeterminate-length array"
and the end-of-file assertion from "check that one element has been
produced" to "check we're still in the indeterminata-length list".

> What if we changed *everything* to a CBOR sequence? (i.e., save one
> byte at the top-level and replace every array of nested elements/form
> fields with a `bstr .cborseq X`)

That'd be hard on the enoder unless it chunks it up all the time. It's
hard enough to know the number of records beforehand (and for that,
indeterminate-length encoding is much easier to do in arrays than in
strings). Knowing the number of bytes in a CBOR tree is already
nontrivial when the structure is known (eg. in OSCORE where 3 levels of
nesting are tracked at some point), and would be quite a hassle here.

**Comment by ektrah** [comment]
> That'd be hard on the enoder unless it chunks it up all the time.

Agreed. Also, patching in the number of bytes using a variable length encoding is a pain.

> indeterminate-length encoding is much easier 

We require Core Deterministic Encoding, so no indeterminate-length encoding in CoRAL.

**Comment by cabo** [comment]
On 2020-01-14, at 22:06, ektrah <notifications@github.com> wrote:
> 
> indeterminate-length encoding is much easier
> 
> We require Core Deterministic Encoding, so no indeterminate-length encoding in CoRAL.

Yes, but that is not limiting the outer CBOR Sequence.

Grüße, Carsten
